akita/src/api/entity-store.ts

```diff
@@ -125,23 +125,37 @@ export class EntityStore<S extends EntityState<E>, E> extends Store<S> {
    *   name: 'New Name'
    * });
    *
+   * this.store.update(e => e.name === 'value', {
+   *   name: 'New Name'
+   * });
    *
    * this.store.update(null, {
    *   name: 'New Name'
    * });
    *
    */
   update(id: ID | ID[] | null, newStateFn: ((entity: Readonly<E>) => Partial<E>));
-  update(id: ID | ID[] | null, newStateFn: Partial<E>);
+  update(id: ID | ID[] | null, newState: Partial<E>);
   update(id: ID | ID[] | null, newState: Partial<S>);
+  update(predicate: ((entity: Readonly<E>) => boolean), newStateFn: ((entity: Readonly<E>) => Partial<E>));
+  update(predicate: ((entity: Readonly<E>) => boolean), newState: Partial<E>);
+  update(predicate: ((entity: Readonly<E>) => boolean), newState: Partial<S>);
   update(newState: Partial<S>);
-  update(stateOrId: ID | ID[] | null | Partial<S>, newStateFn?: ((entity: Readonly<E>) => Partial<E>) | Partial<E> | Partial<S>) {
-    const ids = toBoolean(stateOrId) ? coerceArray(stateOrId) : this._value().ids;
-    isDev() && globalState.setAction({ type: 'Update Entity', entityId: ids });
-
+  update(idsOrFn: ID | ID[] | null | Partial<S> | ((entity: Readonly<E>) => boolean), newStateOrFn?: ((entity: Readonly<E>) => Partial<E>) | Partial<E> | Partial<S>) {
     this.setState(state => {
-      const newState = isFunction(newStateFn) ? newStateFn(state.entities[stateOrId as ID]) : newStateFn;
-      return _crud._update(state, ids, newState);
+      let ids: ID[] = [];
+      if (isFunction(idsOrFn)) {
+        for (const id in state.entities) {
+          if (idsOrFn(state.entities[id])) {
+            ids.push(id);
+          }
+        }
+      } else {
+        ids = toBoolean(idsOrFn) ? coerceArray(idsOrFn) : this._value().ids;
+      }
+      isDev() && globalState.setAction({ type: 'Update Entity', entityId: ids });
+
```

Comment:
this one is an interesting case. Please correct me if my understanding is off. The desired behavior is that if no entities are found for the predicate then we should not update the store. Now, what does it actually mean?

Current workflow: `_crud._update(state, ids, newStateOrFn)` returns an immutably updated state, which is then used by `setState()` function which propagates this new state value to BehaviorSubject.

So if `_crud._update` is called with empty `ids` array, it will still result in new value pushed to BehaviorSubject. If we want to prevent that then, we'd have to make a change to `setState()` function. Something like:
```ts
// entity-store.ts; update function
...
if (!ids.length) return; // we return undefined value, which we'll check for in setState method
return _crud._update(state, ids, newStateOrFn);

// store.ts; setState function
const newStoreValue = newStateFn(this._value());
if (!newStoreValue) return;
this.storeValue = __DEV__ ? deepFreeze(newStoreValue) : newStoreValue;
...
```

The same logic would apply to remove method.
Does this sound reasonalble to you? The change feels a bit weird (I'm not quite sure about returning undefined in the `setState` callback) so I wanted to run this by you before making this change

Reply:
If we didn't have ids, why we should update the state? There is nothing to update, so we don't need to call `_crud._update`.

Reply:
Check this line:
```
if (this._value().ids.length === 0) return;
```

Reply:
This code exists for a case when someone calls `remove()`, but the store is empty. Updating the store is redundant in this case.

Reply:
I understand that. The `if (this._value().ids.length === 0) return;` cide is outside of the `setStore()` function.
In our case, we determine if there are ids to update/remove inside the `setStore` function's callback because we need to use the `state` variable. So the `return;` would be in a different context.

Reply:
I suppose I could move the logic that determines ids outside the `setState()` callback if I use `this.entities` instead of `state.entities`. I'm not sure if there are any implications to doing that

Reply:
You can do the ids calc outside the setState function. Use `this._value()` - it's the same pointer that you get inside the `setState` callback.

Reply:
awesome, thanks